Improve js_of_ocaml support

hhugo opened the pull request:
Please review commit by commit.

- [x] the tests stanza should run tests for all the modes defined. Allowing running tests in javascript automatically (fix #4951)
- [x] configure the default alias for runtest with js_of_ocaml. The main motivation is that we want people to be able to run tests even without node installed (fix #4999)
- [x] configure compilation mode in env (fix #970)
- [x] configure jsoo flags in env (fix #1613)
- [x] make js build work with disable_dynamically_linked_foreign_archives=true
- [x] pass "-no-check-prims" when building bytecode for jsoo (fix #4027)
- [x] add/fix support for build-info+jsoo (fix #4444)
- [x] add tests
- [x] udpate documentation

Extends https://github.com/ocaml/dune/pull/1044

hhugo commented on src/dune_rules/test_rules.ml:
```diff
@@ -17,58 +17,93 @@ let rules (t : Dune_file.Tests.t) ~sctx ~dir ~scope ~expander ~dir_contents =
   let open Memo.Build.O in
+  let modes =
+    if Dune_project.dune_version (Scope.project scope) < (3, 0) then
+      [ (Dune_file.Executables.Link_mode.exe, Loc.none) ]
```
I refactored the code a bit, not needed anymore

hhugo commented:
@nojb, I'm not planning to fix #4213 in the PR. sorry.

hhugo commented:
@rgrinberg, I think this is ready for another round of review !

hhugo commented:
@nojb, I changed by mind, the PR also fixes #4213

hhugo commented on test/blackbox-tests/test-cases/jsoo/tests.t/run.t:
```diff
@@ -1,8 +1,6 @@
-  File "run.t", line 1, characters 0-0:
-  Error: Files _build/default/run.t and _build/default/run.t.corrected differ.
+  $ dune build @runtest-js
+  Error: Alias "runtest-js" specified on the command line is empty.
```
@rgrinberg, @jeremiedimino how can I convince dune that this alias exists ?

jeremiedimino replied:
You can make it a "standard" alias. See function `make_standard` in `src/dune_engine/alias.ml`.

hhugo replied:
I want to have the alias configurable in env by users,  does it make sense to call make_standard ?

jeremiedimino replied:
Standard only means that if the alias is empty then Dune doesn't complain about it.

hhugo replied:
Thanks

hhugo commented:
I don't understand some diffs when running tests. Any idea ?
```diff
$ make test
./dune.exe runtest
File "test/blackbox-tests/test-cases/exec-missing.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/exec-missing.t/run.t
++++++ test/blackbox-tests/test-cases/exec-missing.t/run.t.corrected
File "test/blackbox-tests/test-cases/exec-missing.t/run.t", line 8, characters 0-1:
 |When using dune exec, the external-lib-deps command refers to the executable:
 |
 |  $ dune exec ./x.exe
 |  File "dune", line 3, characters 12-26:
 |  3 |  (libraries does-not-exist))
 |                  ^^^^^^^^^^^^^^
 |  Error: Library "does-not-exist" not found.
-|  -> required by _build/default/x.exe
 |  [1]
File "test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t
++++++ test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t.corrected
File "test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t", line 5, characters 0-1:
 |Test that `enabled_if` fields work as expected.
 |
 |This alias is disabled, building it should do nothing:
 |  $ dune build @x
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |This one is enabled:
 |  $ dune build @y
-|  Building alias y
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |This rule is disabled, trying to build a should fail:
 |  $ dune build a
-|  Error: Don't know how to build a
-|  Hint: did you mean b?
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
 |  [1]
 |
 |This one is enabled:
 |  $ dune build b
-|  Building file b
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |Test the enabled_if field for libraries:
 |
 |  $ dune build foo
-|  Error: Don't know how to build foo
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
 |  [1]
 |
 |  $ dune build main.exe
 |  File "dune", line 35, characters 12-15:
 |  35 |  (libraries foo))
 |                   ^^^
 |  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
 |  -> required by library "bar" in _build/default
 |  -> required by executable main in dune:44
-|  -> required by _build/default/main.exe
 |  [1]
File "test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t
++++++ test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t.corrected
File "test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t", line 10, characters 0-1:
 |Executable that tries to use two implementations for the same virtual lib
 |  $ dune build
 |  Error: Conflicting implementations for virtual library "vlib" in
 |  _build/default/vlib:
 |  - "impl1" in _build/default/impl1
 |    -> required by library "bar" in _build/default
 |  - "impl2" in _build/default/impl2
 |  This cannot work.
 |  -> required by executable foo in dune:2
-|  -> required by _build/default/foo.exe
-|  -> required by alias default in dune:11
 |  [1]
File "test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t
++++++ test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t.corrected
File "test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t", line 6, characters 0-1:
 |Executable that tries to build against a virtual library without an implementation
 |  $ dune build
 |  Error: No implementation found for virtual library "vlib" in
 |  _build/default/vlib.
 |  -> required by executable foo in dune:2
-|  -> required by _build/default/foo.exe
-|  -> required by alias default in dune:5
 |  [1]
make: *** [Makefile:75: test] Error 1
```

hhugo commented:
> I don't understand some diffs when running tests. Any idea ?

Reply to myself, fixed by https://github.com/ocaml/dune/pull/5049/commits/64517a008c00aea15e529e19f2334145722696a3

hhugo commented on src/dune_rules/exe.ml:
```diff
@@ -21,7 +21,11 @@ module Linkage = struct
+  let byte_for_jsoo =
+    { mode = Byte_for_jsoo
+    ; ext = ".bc-for-jsoo"
+    ; flags = [ "-no-check-prims" ]
```
Should this be guarded by a check on version ?

hhugo commented:
I've cleaned up the history. It should be fully ready for review.

hhugo commented:
I'm not sure what to make of the macos failure
```diff
diff --git a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
index c049768..b33149d 100644
--- a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml
+++ b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
@@ -41,12 +41,13 @@ let%expect_test _ =
   print_events 1;
   [%expect
     {|
-    { path = In_source_tree "."; kind = "Created" }
-    { path = In_source_tree "x"; kind = "Unknown" }
-    Got more events than expected: expected 1, saw 2 |}];
+    { path = In_build_dir "."; kind = "Created" } |}];
   Unix.rename "x" "y";
   print_events 0;
   [%expect {|
+    { path = In_source_tree "x"; kind = "Unknown" }
+    { path = In_source_tree "y"; kind = "Unknown" }
+    Got more events than expected: expected 0, saw 2
 |}];
   let (_ : _) = Fpath.mkdir_p "d/w" in
   Stdio.Out_channel.write_all "d/w/x" ~data:"x";
@@ -56,10 +57,8 @@ let%expect_test _ =
   print_events 1;
   [%expect
     {|
-    { path = In_source_tree "x"; kind = "Unknown" }
-    { path = In_source_tree "y"; kind = "Unknown" }
     { path = In_source_tree "d"; kind = "Created" }
     { path = In_source_tree "d/w"; kind = "Created" }
     { path = In_source_tree "d/w/x"; kind = "Unknown" }
     { path = In_source_tree "d/w/y"; kind = "Unknown" }
-    Got more events than expected: expected 1, saw 6 |}]
+    Got more events than expected: expected 1, saw 4 |}]
make: *** [test] Error 1
'make test' failed with code 2
```

rgrinberg commented:
You can ignore the mac error for now. I'm fixing it in another PR.
On Nov 5, 2021, 8:47 AM -0600, hhugo ***@***.***>, wrote:
> I'm not sure what to make of the macos failure
> diff --git a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
> index c049768..b33149d 100644
> --- a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml
> +++ b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
> @@ -41,12 +41,13 @@ let%expect_test _ =
>   print_events 1;
>   [%expect
>     {|
> -    { path = In_source_tree "."; kind = "Created" }
> -    { path = In_source_tree "x"; kind = "Unknown" }
> -    Got more events than expected: expected 1, saw 2 |}];
> +    { path = In_build_dir "."; kind = "Created" } |}];
>   Unix.rename "x" "y";
>   print_events 0;
>   [%expect {|
> +    { path = In_source_tree "x"; kind = "Unknown" }
> +    { path = In_source_tree "y"; kind = "Unknown" }
> +    Got more events than expected: expected 0, saw 2
> |}];
>   let (_ : _) = Fpath.mkdir_p "d/w" in
>   Stdio.Out_channel.write_all "d/w/x" ~data:"x";
> @@ -56,10 +57,8 @@ let%expect_test _ =
>   print_events 1;
>   [%expect
>     {|
> -    { path = In_source_tree "x"; kind = "Unknown" }
> -    { path = In_source_tree "y"; kind = "Unknown" }
>     { path = In_source_tree "d"; kind = "Created" }
>     { path = In_source_tree "d/w"; kind = "Created" }
>     { path = In_source_tree "d/w/x"; kind = "Unknown" }
>     { path = In_source_tree "d/w/y"; kind = "Unknown" }
> -    Got more events than expected: expected 1, saw 6 |}]
> +    Got more events than expected: expected 1, saw 4 |}]
> make: *** [test] Error 1
> 'make test' failed with code 2
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ggreif commented:
@hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?

hhugo commented:
> @hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?

No, you can get rid of it

ggreif commented:
> > @hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?
> 
> No, you can get rid of it

#6923

ggreif commented:
@hhugo I have trouble getting this working:
```
File "js/dune", line 22, characters 53-63:
22 |  (js_of_ocaml (flags <flags>)(build_runtime <flags>)(link_flags <flags>))
                                                          ^^^^^^^^^^
Error: Unknown field link_flags
```
I am using `dune` v3.2. The line 22 above is straight from the manual. (Forget the `<flags>` placeholders for now, they are irrelevant.)

hhugo commented:
> @hhugo I have trouble getting this working:
> 
> ```
> File "js/dune", line 22, characters 53-63:
> 22 |  (js_of_ocaml (flags <flags>)(build_runtime <flags>)(link_flags <flags>))
>                                                           ^^^^^^^^^^
> Error: Unknown field link_flags
> ```
> 
> I am using `dune` v3.2. The line 22 above is straight from the manual. (Forget the `<flags>` placeholders for now, they are irrelevant.)

What `dune lang` are you using ?

ggreif commented:
Yes, `echo '(lang dune 3.0)' > js/dune-project` seems to solve my problem! Thanks!
